generate/error_code_generator.go

```diff
@@ -0,0 +1,190 @@
+
+	decoder := yaml.NewDecoder(yamlFile)
+
+	var values []interface{}
+
+	for {
+		var value map[string]interface{}
+		if err := decoder.Decode(&value); err != nil {
+			if err == io.EOF {
+				break
+			}
+			panic(fmt.Errorf("document decode failed"))
+		}
+		values = append(values, value)
+	}
+
+	errors := make([]ErrorType, 0)
+
+	for i := 0; i < len(values); i++ {
+		entry := values[i].(map[string]interface{})
+
+		errorType, ok := entry["type"].(string)
+		if !ok {
+			panic(fmt.Sprintf("Error entry at position %d is missing the type attribute", i))
+		}
+		for _, keyName := range requiredErrorKeys {
+			if _, ok := entry[keyName]; !ok {
+				panic(fmt.Sprintf("Parameter entry '%s' is missing required key '%s'", errorType, keyName))
+			}
+		}
+		camelScopeName := handleCaseConversion(errorType)
+		scopeNameInSnake := strings.Replace(camelScopeName, ".", "_", 1)
+		r := []rune(scopeNameInSnake)
+		r[0] = unicode.ToUpper(r[0])
+		displayName := string(r)
+
+		exitCode, ok := entry["clientExitCode"].(int)
+		if !ok {
+			panic(fmt.Sprintf("Error entry at position %d is missing the exit code attribute", i))
+		}
+
+		code, ok := entry["code"].(int)
+		if !ok {
+			panic(fmt.Sprintf("Error entry at position %d is missing the code attribute", i))
+		}
+
+		retryable, ok := entry["retryable"].(bool)
+		if !ok {
+			panic(fmt.Sprintf("Error entry at position %d is missing the retryable attribute", i))
+		}
+
+		errors = append(errors, ErrorType{Raw: errorType, Display: displayName, ExitCode: exitCode,
+			Code: code, Retryable: retryable})
+	}
+
+	// Create the fike to be generated
+	f, err := os.Create("../error_codes/error_codes.go")
+	if err != nil {
+		panic(err)
+	}
+	defer f.Close()
+
+	err = errorTemplate.Execute(f, struct {
+		PelicanErrors []ErrorType
+	}{
+		PelicanErrors: errors,
+	})
+
+	if err != nil {
+		panic(err)
+	}
+}
+
+var errorTemplate = template.Must(template.New("").Parse(`// Code generated by go generate; DO NOT EDIT.
+/***************************************************************
+ *
+ * Copyright (C) 2024, Pelican Project, Morgridge Institute for Research
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you
+ * may not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ ***************************************************************/
+
+package error_codes
+
+import (
+	"fmt"
+)
+
+type PelicanError struct {
+	errorType string
+	exitCode  int
+	code      int
+	retryable bool
+	err       error
+}
+
+{{- range $idx, $pelicanError := .PelicanErrors}}
```

Comment:
The beginning `-` you have here and in the `end` statements are why you're not getting newlines between functions in the generated code. It specifies "remove all whitespace on the side of the statement that has -"